rustls/src/client/ech.rs

```diff
@@ -109,3 +136,364 @@ impl EchConfig {
+            .map(|ext| ext.ext_type())
+            .collect();
+
+        // Set the inner hello random to the one we generated when creating the ECH state.
+        // We hold on to the inner_hello_random in the ECH state to use later for confirming
+        // whether ECH was accepted or not.
+        inner_hello.random = self.inner_hello_random;
+
+        // If we're resuming, we need to update the PSK binder in the inner hello.
+        if let Some(resuming) = resuming.as_ref() {
+            let mut chp = HandshakeMessagePayload {
+                typ: HandshakeType::ClientHello,
+                payload: HandshakePayload::ClientHello(inner_hello),
+            };
+
+            // Retain the early key schedule we get from processing the binder.
+            self.early_data_key_schedule = Some(tls13::fill_in_psk_binder(
+                resuming,
+                &self.inner_hello_transcript,
+                &mut chp,
+            ));
+
+            // fill_in_psk_binder works on an owned HandshakeMessagePayload, so we need to
+            // extract our inner hello back out of it to retain ownership.
+            inner_hello = match chp.payload {
+                HandshakePayload::ClientHello(chp) => chp,
+                // Safety: we construct the HMP above and know its type unconditionally.
+                _ => unreachable!(),
+            };
+        }
+
+        // Repeating large extensions between ClientHelloInner and ClientHelloOuter can lead to excessive
+        // size. To reduce the size impact, the client MAY substitute extensions which it knows will be
+        // duplicated in ClientHelloOuter.
+
+        // TODO(@cpu): Extension compression would be handled here-ish.
+
+        // 5.1 "Encoding the ClientHelloInner"
+
+        // Setting the legacy_session_id field to the empty string.
+        // Preserve these for reuse
+        let original_session_id = inner_hello.session_id;
+
+        // SessionID is required to be empty in the EncodedClientHelloInner.
+        inner_hello.session_id = SessionId::empty();
+
+        // Encode the inner hello with the empty session ID.
+        let mut encoded_hello = inner_hello.get_encoding();
+
+        // Restore session ID.
+        inner_hello.session_id = original_session_id;
+
+        trace!("ECH Inner Hello: {:#?}", inner_hello);
+
+        // Calculate padding
+        // max_name_len = L
+        let max_name_len = self.maximum_name_length;
+        let max_name_len = if max_name_len > 0 { max_name_len } else { 255 };
+
+        let padding_len = match &self.inner_name {
+            ServerName::DnsName(name) => {
+                // name.len() = D
+                // max(0, L - D)
+                core::cmp::max(0_u8, max_name_len.saturating_sub(name.as_ref().len() as u8))
+            }
+            _ => {
+                // L + 9
+                // "This is the length of a "server_name" extension with an L-byte name."
+                max_name_len + 9
```

Comment:
I think `max_name_len` and `padding_len` are both u8 at this point (could be wrong) so `max_name_len + 9` could certainly overflow based on DNS  EchConfig contents. widen `padding_len` to usize?

Reply:
Very good catch, thank you. We were promoting `padding_len` to usize later on, so I shifted that to happen here to avoid the overflow and left a comment about it.